src/main.rs

```diff
@@ -40,22 +39,56 @@ fn get_user_agent() -> &'static str {
     }
 }
 
-#[exit_status::main]
-fn main() -> Result<i32> {
+fn main() {
     let args = Cli::from_args();
 
+    let bin_name = args.bin_name.clone();
+    let url = args.url.clone();
+    let native_tls = args.native_tls;
+
+    match inner_main(args) {
+        Ok(exit_code) => {
+            process::exit(exit_code);
+        }
+        Err(err) => {
+            eprintln!("{}: error: {:?}", bin_name, err);
+            if !native_tls && err.root_cause().to_string() == "invalid dnsname" {
+                eprintln!();
+                if utils::requires_native_tls(&url) {
+                    eprintln!("rustls does not support HTTPS for IP addresses.");
+                } else {
+                    // Maybe we went to https://<IP> after a redirect?
+                    eprintln!(
+                        "This may happen because rustls does not support HTTPS for IP addresses."
+                    );
+                }
+                if cfg!(feature = "native-tls") {
+                    eprintln!("Try using the --native-tls flag.");
+                } else {
+                    eprintln!("Consider building with the `native-tls` feature enabled.");
+                }
+            }
+            process::exit(1);
+        }
+    }
+}
+
+fn inner_main(args: Cli) -> Result<i32> {
     if args.curl {
         to_curl::print_curl_translation(args)?;
         return Ok(0);
     }
 
-    let request_items = RequestItems::new(args.request_items);
-    let query = request_items.query();
-    let (headers, headers_to_unset) = request_items.headers()?;
-    let url = construct_url(&args.url, args.default_scheme.as_deref(), query)?;
+    let warn = {
+        let bin_name = &args.bin_name;
+        move |msg| eprintln!("{}: warning: {}", bin_name, msg)
+    };
```

Comment:
Would it suffice to just use `CARGO_PKG_NAME` instead of `args.bin_name`? This would allow to turn this into a macro that is similar to `println()`.

```rust
warn(&format!("HTTP {}", status));
// vs
warn!("HTTP {}", status)
```

Reply:
My thinking was that you want to see which command caused the message, and the name it was invoked with is more informative.

It actually could be made a macro but I feel macros should be a tool of last resort. If we have a lot of warnings that require string formatting then it could be worth it but a closure is good enough for now.